### PR TITLE
thunderscope: dynamic parameters protobuf code generation

### PR DESCRIPTION
**Pull request by akhilveeraghanta**
<!---
This file outlines a list of common things that should be addressed when opening a PR. It's built from previous issues we've seen in a lot of pull requests. If you notice something that's being noted in a lot of PR's, it should probably be added here to help save people time in the future.
-->

## Please fill out the following before requesting review on this PR

### Description
* generates protobufs for dynamic paremeters
* generates toProto and loadFromProto on the c++ function
* test for that
<!--
    Give a high-level description of the changes in this PR
-->

### Testing Done
unit test
<!--
    Outline any testing that was done for these changes. This could be unit tests, integration tests,etc.
-->

### Resolved Issues
N/A
<!--
    Link any issues that this PR resolved. Ex. `resolves #1, #2, and #5` (note that they MUST be specified like this so Github can automatically close them then this PR merges)
-->

Sample:

~~~proto
/**                                                                                                                                                                                                                                                                                                                                                                          
 *  !! WARNING !!                                                                                                                                                                                                                                                                                                                                                            
 *                                                                                                                                                                                                                                                                                                                                                                           
 *  THIS FILE IS AUTOGENERATED, ANY CHANGES MADE WILL BE LOST                                                                                                                                                                                                                                                                                                                
 *                                                                                                                                                                                                                                                                                                                                                                           
 *  !! WARNING !!                                                                                                                                                                                                                                                                                                                                                            
 */                                                                                                                                                                                                                                                                                                                                                                          
syntax = "proto3";                                                                                                                                                                                                                                                                                                                                                           
import "google/protobuf/any.proto";                                                                                                                                                                                                                                                                                                                                          
package TbotsProto;                                                                                                                                                                                                                                                                                                                                                          
message AiConfig {                                                                                                                                                                                                                                                                                                                                                           
google.protobuf.Any corner_kick_play_config = 1;                                                                                                                                                                                                                                                                                                                             
google.protobuf.Any goalie_tactic_config = 2;                                                                                                                                                                                                                                                                                                                                
google.protobuf.Any navigator_config = 3;                                                                                                                                                                                                                                                                                                                                    
google.protobuf.Any passing_config = 4;                                                                                                                                                                                                                                                                                                                                      
google.protobuf.Any robot_navigation_obstacle_config = 5;                                                                                                                                                                                                                                                                                                                    
google.protobuf.Any attacker_tactic_config = 6;                                                                                                                                                                                                                                                                                                                              
google.protobuf.Any shoot_or_pass_play_config = 7;                                                                                                                                                                                                                                                                                                                           
}                                                                                                                                                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                                                                                                                                             
message SslSimulationConfig {                                                                                                                                                                                                                                                                                                                                                
int32 team_channel = 1;                                                                                                                                                                                                                                                                                                                                                      
string network_interface = 2;                                                                                                                                                                                                                                                                                                                                                
string ssl_simulator_ip_address = 3;                                                                                                                                                                                                                                                                                                                                         
int32 ssl_simulator_port = 4;                                                                                                                                                                                                                                                                                                                                                
}                                                                                                                                                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                                                                                                                                             
message AiControlConfig {                                                                                                                                                                                                                                                                                                                                                    
bool run_ai = 1;                                                                                                                                                                                                                                                                                                                                                             
bool override_ai_play = 2;                                                                                                                                                                                                                                                                                                                                                   
string current_ai_play = 3;                                                                                                                                                                                                                                                                                                                                                  
}                                                                                                                                                                                                                                                                                                                                                                            
                                                                                                                                                                                                                                                                                                                                                                             
message StandaloneSimulatorConfig {                                                                                                                                                                                                                                                                                                                                          
int32 yellow_team_channel = 1;                                                                                                                                                                                                                                                                                                                                               
int32 blue_team_channel = 2;                                                                                                                                                                                                                                                                                                                                                 
string network_interface = 3;                                                                                                                                                                                                                                                                                                                                                
string vision_ipv4_address = 4;                                                                                                                                                                                                                                                                                                                                              
int32 vision_port = 5;                                                                                                                                                                                                                                                                                                                                                       
}      
~~~

### Length Justification and Key Files to Review
N/A
<!-- If this pull request is longer then **500** lines (additions + deletions), please justify here why we *cannot* break this up into multiple pull requests
and list the key files that contain the main idea of your PR -->

### Review Checklist

<!--
    (Please check every item to indicate your code complies with it (by changing `[ ]`->`[x]`). This will hopefully save both you and the reviewer(s) a lot of time!)
-->

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

<!--
    Feel free to make additions of things that we should be checking to this file if you think there's something missing!!!!
    At the same time, consider that adding things to this list increases the burden on everyone opening a pull request. 
    Perhaps there is a way we can automatically enforce whatever item you want to add?
-->

### CODE REVIEW

File: src/shared/parameter/generation_scripts/cpp_header_config.py [hunk 1]
```diff
@@ -5,13 +5,15 @@
 from case_conversion import to_snake_case
 import networkx as nx
 
+TAB_SIZE = 4
+HALF_TAB_SIZE = 2
+INDENT_ONCE = 1
+INDENT_TWICE = 2
```
Comment (jonathanlew) [inline src/shared/parameter/generation_scripts/cpp_header_config.py:11]: :exploding_head:

File: src/shared/parameter/generation_scripts/cpp_header_config.py [hunk 2]
```diff
@@ -61,6 +63,10 @@
 
     const ParameterList& getParameterList() const;
 
+    TbotsProto::{config_name} toProto() const;
+
+    void loadFromProto(const TbotsProto::{config_name}& config_proto);
```
Comment (jonathanlew) [inline src/shared/parameter/generation_scripts/cpp_header_config.py:68]: wondering if we should have function comments

File: src/shared/parameter/cpp_config_test.cpp [hunk 3]
```diff
@@ -461,13 +461,39 @@ class TestParameterMutation : public YamlLoadFixture
...
+    // Create a new top level config
+    const std::shared_ptr<ThunderbotsConfig> mutable_dynamic_param_loaded_from_proto =
+        std::make_shared<ThunderbotsConfig>();
 
-    mutate_all_parameters(MutableDynamicParameters);
-    assert_mutation(DynamicParameters, config_yaml);
+    // Load top level config from the proto and make sure that the mutations across all
+    // configs are reflected
+    mutable_dynamic_param_loaded_from_proto->loadFromProto(proto);
+    assert_mutation(mutable_dynamic_param_loaded_from_proto, config_yaml);
```
Comment (jonathanlew) [inline src/shared/parameter/cpp_config_test.cpp]: wondering why these functions are snake_case lol

**Review by jonathanlew — COMMENTED** [review]
nice work

File: src/shared/parameter/generation_scripts/cpp_header_config.py (hunk 2 above)
Comment (akhilveeraghanta) [reply src/shared/parameter/generation_scripts/cpp_header_config.py:68]: hmm, i don't think anyone really looks at the autogenerated code

I think #1738 should document this

File: src/shared/parameter/generation_scripts/cpp_header_config.py (hunk 1 above)
Comment (akhilveeraghanta) [reply src/shared/parameter/generation_scripts/cpp_header_config.py:11]: thought it was more readable than a random 1/2 scattered about

File: src/shared/parameter/cpp_config_test.cpp (hunk 3 above)
Comment (akhilveeraghanta) [reply src/shared/parameter/cpp_config_test.cpp]: not sure lol, `snake_case` is superior but ill change it to camelCase for consistency sake

**Review by jonathanlew — APPROVED** [review]
LGTM!

**Review by JoshWong16 — APPROVED** [review]
Nice! Really helpful for scaling out Thunderscope.

**Pull request merged**